### ISSUE TITLE
Dash for R API does not currently allow user-defined routes

### ISSUE THREAD

**Issue by rpkyle** [plotly/dashR#131]
As described in https://dash.plot.ly/integrating-dash and seen in [code examples](https://community.plot.ly/t/allowing-users-to-download-csv-on-click/5550/9) posted to our community forum, Dash for Python permits app developers to leverage Flask support for adding new routes. 

Dash for R currently lacks this functionality, so Dash for Python code like this:

```
@app.callback(Output('my-link', 'href'), [Input('my-dropdown', 'value')])
def update_link(value):
    return '/dash/urlToDownload?value={}'.format(value)

@app.server.route('/dash/urlToDownload') 
def download_csv():
    value = flask.request.args.get('value')
    # create a dynamic csv or file here using `StringIO` 
    # (instead of writing to the file system)
    strIO = StringIO.StringIO()
    strIO.write('You have selected {}'.format(value)) 
    strIO.seek(0)    
    return send_file(strIO,
                     mimetype='text/csv',
                     attachment_filename='downloadFile.csv',
                     as_attachment=True)
```

currently has no direct counterpart on the R side. We should expose the ability for application developers to define new routes within the Dash for R API. One possible option is through the use of `fiery` plugins. e.g. a `server_route` method could be added which exposes the ability to add plugins with new route handlers:

```
plugin <- list(
    on_attach = function(server) {
     router <- server$plugins$request_routr
     route <- Route$new()
     route$add_handler('all', '*', function(request, response, arg_list, ...) {
       message('Hello')
       TRUE
    })
    router$add_route(route, 1)
    },
  name = 'Hello_plugin',
  require = 'request_routr'
)
```

Thanks to @CanerIrfanoglu for raising this issue initially.

**Comment by rpkyle** [comment]
A `server_route` method would greatly simplify redirects, as recently required by edits to the Dash for R documentation. Adding these is currently possible, but requires far more verbose and user-unfriendly code, e.g.

```
plugin <- list(
  on_attach = function(server) {
    router <- server$plugins$request_routr
    route <- routr::Route$new()
    redirect_getting_started <- function(request, response, keys, ...) {
      response$status <- 301L
      response$set_header('Location', '/getting-started')
      TRUE
    }
    redirect_getting_started_2 <- function(request, response, keys, ...) {
      response$status <- 301L
      response$set_header('Location', '/getting-started-part-2')
      TRUE
    }
    redirect_state <- function(request, response, keys, ...) {
      response$status <- 301L
      response$set_header('Location', '/state')
      TRUE
    }
    route$add_handler('get', '/layout', redirect_getting_started)
    route$add_handler('get', '/basic-callbacks', redirect_getting_started_2)
    route$add_handler('get', '/basic-callbacks', redirect_state)
    router$add_route(route, "redirects")
  },
  name = 'redirect_urls',
  require = 'request_routr'
)

app$server$attach(plugin)
```

**Comment by rpkyle** [comment]
Closed by #225.
